### PR TITLE
feat(storage): Add support for if_(meta)generation_(not_)match to File operations

### PR DESCRIPTION
**Pull request by quartzmo**
- [x] Add `if_(meta)generation_match` options to `Bucket#compose`
- [x] Add `if_(meta)generation_(not_)match` options to `Bucket#create_file`
- [x] Add `if_(meta)generation_(not_)match` options to `Bucket#file`
- [x] Add `if_(meta)generation_(not_)match` options to `File#rewrite`
- [x] Add `generation` and `if_(meta)generation_(not_)match` options to `File#update`
- [x] Add `generation` and `if_(meta)generation_(not_)match` options to `File::Acl` predefined_acl methods

closes: #11259

### CODE REVIEW

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 1]
```diff
@@ -1061,14 +1074,25 @@ def files prefix: nil, delimiter: nil, token: nil, max: nil,
         #   file = bucket.file "path/to/my-file.ext"
         #   puts file.name
         #
-        def file path, generation: nil, skip_lookup: nil, encryption_key: nil
+        def file path,
+                 generation: nil,
+                 if_generation_match: nil,
```
Comment (frankyn) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1079]: Do these parameters get propagated across all Object operations that support these preconditions?
Similar question for bucket.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 2]
```diff
@@ -1334,6 +1393,12 @@ def create_file file, path = nil, acl: nil, cache_control: nil,
         #   used. All source files must have been encrypted with the same key,
         #   and the resulting destination file will also be encrypted with the
         #   key.
+        # @param [Integer] if_generation_match Makes the operation conditional
```
Comment (frankyn) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1396]: I would call out that this parameter is for the destination object name.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 3]
```diff
@@ -1382,7 +1447,12 @@ def create_file file, path = nil, acl: nil, cache_control: nil,
         #
         #   new_file = bucket.compose [file_1, file_2], "path/to/new-file.ext"
         #
-        def compose sources, destination, acl: nil, encryption_key: nil
+        def compose sources,
```
Comment (frankyn) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1450]: How do sources pass in if(_*meta)_generation?

**Review by frankyn — CHANGES_REQUESTED** [review]
Thanks @quartzmo!

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 1 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1079]: No. Is that the desired behavior? To store the precondition values in the `File` instance and propagate them to `rewrite`, `update`, etc? It sounds a bit tricky and would require additional docs to clarify the behind-the-scenes behavior, but I can attempt it.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 3 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1450]: Thanks for catching this. I'll expose options to set the `Google::Apis::StorageV1::ComposeRequest::SourceObject::ObjectPreconditions`.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 2 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1398]: Done, thanks!

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 3 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1453]: I added a `Google::Cloud::Storage::File::Reference` value object to accept the `if_generation_match` values for sources. Please let me know what you think about this approach, including the naming, thanks.

File: google-cloud-storage/lib/google/cloud/storage/file.rb [hunk 4]
```diff
@@ -1849,6 +1979,83 @@ def self.gapi_from_attrs gapi, attributes
           Google::Apis::StorageV1::Object.new(**attr_params)
         end
 
+        ##
+        # Represents a reference to a file, with optional preconditions.
+        # See `sources` in {Bucket#compose}.
```
Comment (frankyn) [inline google-cloud-storage/lib/google/cloud/storage/file.rb:1984]: @quartzmo, recommendation: please split out Bucket#compose preconditions support into a separate follow-up FR. I suspect we will need to dig in a bit more on what the best way to represent this information will be. I don't want to jump to a specific design yet.

**Review by frankyn — CHANGES_REQUESTED** [review]
Thanks @quartzmo, at this point it will be good to leave Bucket#compose for a follow-up issue. Could you file a tracking issue for this? 

Other than that, this LGTM.

File: google-cloud-storage/lib/google/cloud/storage/file.rb (hunk 4 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/file.rb]: Removed Bucket#compose preconditions support from this PR.

**Review by frankyn — APPROVED** [review]
LGTM!

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 3 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1450]: This discussion has been moved to #11690.

File: google-cloud-storage/lib/google/cloud/storage/file.rb (hunk 4 above)
Comment (quartzmo) [reply google-cloud-storage/lib/google/cloud/storage/file.rb]: Moved to #11690.

**Review by dazuma — APPROVED** [review]
LGTM

**Pull request merged**